Assisted factory generation fails if the factory interface is generated and it implements an interface from another module

This is similar to #395, but only affects generated factory interfaces.  If this code is generated, then the subsequent generation of `AssistedFactoryImpl` will fail.

```kotlin
data class AssistedService @AssistedInject constructor(
  @Assisted val string: String
)
    
@AssistedFactory
interface AssistedFactory : Function1<String, AssistedService>
```

start of the stacktrace:
```
e: java.lang.AssertionError: Classifier descriptor of a type should be of type ClassDescriptor: R
	at org.jetbrains.kotlin.resolve.DescriptorUtils.getClassDescriptorForTypeConstructor(DescriptorUtils.java:390)
	at org.jetbrains.kotlin.resolve.DescriptorUtils.getClassDescriptorForType(DescriptorUtils.java:383)
	at com.squareup.anvil.compiler.internal.CompilerUtilsKt.classDescriptorForType(CompilerUtils.kt:93)
	at com.squareup.anvil.compiler.codegen.dagger.AssistedFactoryGenerator$AssistedFactoryFunction$Companion$toAssistedFactoryFunction$returnTypeLazy$1.invoke(AssistedFactoryGenerator.kt:293)
	at com.squareup.anvil.compiler.codegen.dagger.AssistedFactoryGenerator$AssistedFactoryFunction$Companion$toAssistedFactoryFunction$returnTypeLazy$1.invoke(AssistedFactoryGenerator.kt:291)
	at kotlin.UnsafeLazyImpl.getValue(Lazy.kt:81)
	at com.squareup.anvil.compiler.codegen.dagger.AssistedFactoryGenerator.generateFactoryClass(AssistedFactoryGenerator.kt:106)
	at com.squareup.anvil.compiler.codegen.dagger.AssistedFactoryGenerator.generateCodePrivate(AssistedFactoryGenerator.kt:90)
	at com.squareup.anvil.compiler.codegen.PrivateCodeGenerator.generateCode(PrivateCodeGenerator.kt:21)
	at com.squareup.anvil.compiler.codegen.CodeGenerationExtension.analysisCompleted$generateCode(CodeGenerationExtension.kt:92)
	at com.squareup.anvil.compiler.codegen.CodeGenerationExtension.analysisCompleted(CodeGenerationExtension.kt:113)
```

This is happening because the `fun invoke(p1: P1): R` function is invisible to PSI, and the `FunctionDescriptor` we get is from `Function1`, before the types are defined.

--- a/compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/ClassReference.kt
+++ b/compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/ClassReference.kt
@@ -1,0 +1,146 @@
+package com.squareup.anvil.compiler.internal
+
+import com.squareup.anvil.annotations.ExperimentalAnvilApi
+import com.squareup.anvil.compiler.api.AnvilCompilationException
+import com.squareup.anvil.compiler.internal.ClassReference.Descriptor
+import com.squareup.anvil.compiler.internal.ClassReference.Psi
+import org.jetbrains.kotlin.descriptors.ClassDescriptor
+import org.jetbrains.kotlin.descriptors.ModuleDescriptor
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.psi.KtClassOrObject
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameSafe
+
+/**
+ * Used to create a common type between [KtClassOrObject] class references and [ClassDescriptor]
+ * references, to streamline parsing.
+ *
+ * @see allSuperTypeClassReferences
+ * @see requireClassReference
+ */
+@ExperimentalAnvilApi
+public sealed class ClassReference {
+
+  public abstract val fqName: FqName
+
+  public class Psi internal constructor(
+    public val clazz: KtClassOrObject,
+    override val fqName: FqName
+  ) : ClassReference()
+
+  public class Descriptor internal constructor(
+    public val clazz: ClassDescriptor,
+    override val fqName: FqName
+  ) : ClassReference()
+
+  override fun toString(): String {
+    return "${this::class.qualifiedName}($fqName)"
+  }
+
+  override fun equals(other: Any?): Boolean {
+    if (this === other) return true
+    if (other !is ClassReference) return false
+
+    if (fqName != other.fqName) return false
+
+    return true
+  }
+
+  override fun hashCode(): Int {
+    return fqName.hashCode()
+  }
+}
+
+/**
+ * Attempts to resolve the [fqName] to a [ClassDescriptor] first, then falls back to a
+ * [KtClassOrObject] if the descriptor resolution fails. This will happen if the code being parsed
+ * was generated as part of the compilation round for this module.
+ */
+@ExperimentalAnvilApi
+public fun ModuleDescriptor.classReferenceOrNull(fqName: FqName): ClassReference? {
+  return fqName.classDescriptorOrNull(this)?.toClassReference()
+    ?: getKtClassOrObjectOrNull(fqName)?.toClassReference()
+}
+
+@ExperimentalAnvilApi
+public fun ClassDescriptor.toClassReference(): Descriptor {
+  return Descriptor(this, fqNameSafe)
+}
+
+@ExperimentalAnvilApi
+public fun KtClassOrObject.toClassReference(): Psi {
+  return Psi(this, requireFqName())
+}
+
+@ExperimentalAnvilApi
+public fun ModuleDescriptor.requireClassReference(fqName: FqName): ClassReference {
+  return classReferenceOrNull(fqName)
+    ?: throw AnvilCompilationException("Couldn't resolve ClassReference for $fqName")
+}
+
+/**
+ * This will return all super types as [ClassReference], whether they're parsed as [KtClassOrObject]
+ * or [ClassDescriptor]. This will include generated code, assuming it has already been generated.
+ * The returned sequence will be distinct by FqName, and Psi types are preferred over Descriptors.
+ *
+ * The first elements in the returned sequence represent the direct superclass to the receiver. The
+ * last elements represent the types which are furthest up-stream.
+ *
+ * @param includeSelf If true, the receiver class is the first element of the sequence
+ */
+@ExperimentalAnvilApi
+public fun ClassReference.allSuperTypeClassReferences(
+  module: ModuleDescriptor,
+  includeSelf: Boolean = false
+): Sequence<ClassReference> {
+
+  fun KtClassOrObject.superTypeEntryNames() = superTypeListEntries
+    .map { it.requireFqName(module) }
+
+  fun List<FqName>.typeRefs(): Sequence<ClassReference> {
+    return asSequence().map { module.requireClassReference(it) }
+  }
+
+  val initial = sequenceOf(this).drop(if (includeSelf) 0 else 1)
+
+  return generateSequence(initial) { superTypes ->
+    superTypes
+      .mapNotNull { classRef ->
+        when (classRef) {
+          is Psi -> {
+            classRef.clazz.superTypeEntryNames()
+              .takeIf { it.isNotEmpty() }
+              ?.typeRefs()
+          }
+          is Descriptor -> {
+            classRef.clazz.directSuperClassAndInterfaces()
+              .takeIf { it.isNotEmpty() }
+              ?.asSequence()
+              ?.map { module.requireClassReference(it.fqNameSafe) }
+          }
+        }
+      }
+      .flatten()
+      .takeIf { it.firstOrNull() != null }
+  }.flatten()
+    .distinctBy { it.fqName }
+}
+
+/**
+ * @param parameterName The name of the parameter to be found, not including any variance modifiers.
+ * @return The 0-based index of a declared generic type.
+ */
+@ExperimentalAnvilApi
+public fun ClassReference.indexOfTypeParameter(parameterName: String): Int {
+  return when (this) {
+    is Psi -> {
+      clazz
+        .typeParameters
+        .indexOfFirst { it.identifyingElement?.text == parameterName }
+    }
+    is Descriptor -> {
+      clazz
+        .declaredTypeParameters
+        .indexOfFirst { it.name.asString() == parameterName }
+    }
+  }
+}

--- a/compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/CompilerUtils.kt
+++ b/compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/CompilerUtils.kt
@@ -15,7 +15,6 @@ import org.jetbrains.kotlin.name.FqName
 import org.jetbrains.kotlin.name.Name
 import org.jetbrains.kotlin.platform.has
 import org.jetbrains.kotlin.platform.jvm.JvmPlatform
-import org.jetbrains.kotlin.resolve.DescriptorUtils
 import org.jetbrains.kotlin.resolve.constants.ConstantValue
 import org.jetbrains.kotlin.resolve.constants.KClassValue
 import org.jetbrains.kotlin.resolve.constants.KClassValue.Value.NormalClass
@@ -27,6 +26,7 @@ import org.jetbrains.kotlin.resolve.descriptorUtil.getSuperInterfaces
 import org.jetbrains.kotlin.resolve.descriptorUtil.module
 import org.jetbrains.kotlin.types.ErrorType
 import org.jetbrains.kotlin.types.KotlinType
+import org.jetbrains.kotlin.types.TypeUtils
 import org.jetbrains.kotlin.types.typeUtil.supertypes
 import org.jetbrains.org.objectweb.asm.Type
 
@@ -89,8 +89,16 @@ public fun ConstantValue<*>.toType(
 public fun KotlinType.argumentType(): KotlinType = arguments.first().type
 
 @ExperimentalAnvilApi
-public fun KotlinType.classDescriptorForType(): ClassDescriptor {
-  return DescriptorUtils.getClassDescriptorForType(this)
+public fun KotlinType.classDescriptorOrNull(): ClassDescriptor? {
+  return TypeUtils.getClassDescriptor(this)
+}
+
+@ExperimentalAnvilApi
+public fun KotlinType.requireClassDescriptor(): ClassDescriptor {
+  return classDescriptorOrNull()
+    ?: throw AnvilCompilationException(
+      "Unable to resolve type for ${this.asTypeName()}"
+    )
 }
 
 @ExperimentalAnvilApi
@@ -105,7 +113,7 @@ public fun AnnotationDescriptor.scope(module: ModuleDescriptor): ClassDescriptor
       message = "Couldn't find scope for $fqName."
     )
 
-  return annotationValue.argumentType(module).classDescriptorForType()
+  return annotationValue.argumentType(module).requireClassDescriptor()
 }
 
 @ExperimentalAnvilApi
@@ -141,7 +149,7 @@ public fun List<KotlinType>.getAllSuperTypes(): Sequence<FqName> =
     kotlinTypes.ifEmpty { null }?.flatMap { it.supertypes() }
   }
     .flatMap { it.asSequence() }
-    .map { it.classDescriptorForType().fqNameSafe }
+    .map { it.requireClassDescriptor().fqNameSafe }
 
 @ExperimentalAnvilApi
 public fun AnnotationDescriptor.isQualifier(): Boolean {

--- a/compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/FqName.kt
+++ b/compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/FqName.kt
@@ -1,6 +1,7 @@
 package com.squareup.anvil.compiler.internal
 
 import dagger.Lazy
+import dagger.MapKey
 import dagger.Provides
 import javax.inject.Inject
 import javax.inject.Qualifier
@@ -8,6 +9,7 @@ import javax.inject.Qualifier
 internal val jvmSuppressWildcardsFqName = JvmSuppressWildcards::class.fqName
 internal val publishedApiFqName = PublishedApi::class.fqName
 internal val qualifierFqName = Qualifier::class.fqName
+internal val mapKeyFqName = MapKey::class.fqName
 internal val daggerProvidesFqName = Provides::class.fqName
 internal val daggerLazyFqName = Lazy::class.fqName
 internal val injectFqName = Inject::class.fqName

--- a/compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/KotlinPoetUtils.kt
+++ b/compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/KotlinPoetUtils.kt
@@ -266,7 +266,7 @@ public fun KotlinType.asTypeNameOrNull(
 ): TypeName? {
   if (isTypeParameter()) return TypeVariableName(toString())
 
-  val className = classDescriptorForType().asClassName()
+  val className = requireClassDescriptor().asClassName()
   if (!rawTypeFilter(className)) {
     return null
   }
@@ -376,7 +376,7 @@ public fun AnnotationDescriptor.toAnnotationSpec(module: ModuleDescriptor): Anno
       allValueArguments.forEach { (name, value) ->
         when (value) {
           is KClassValue -> {
-            val className = value.argumentType(module).classDescriptorForType()
+            val className = value.argumentType(module).requireClassDescriptor()
               .asClassName()
             addMember("${name.asString()} = %T::class", className)
           }

--- a/compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/PsiUtils.kt
+++ b/compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/PsiUtils.kt
@@ -4,7 +4,6 @@ package com.squareup.anvil.compiler.internal
 
 import com.squareup.anvil.annotations.ExperimentalAnvilApi
 import com.squareup.anvil.compiler.api.AnvilCompilationException
-import com.squareup.anvil.compiler.internal.ClassReference.Descriptor
 import com.squareup.anvil.compiler.internal.ClassReference.Psi
 import com.squareup.kotlinpoet.TypeVariableName
 import org.jetbrains.kotlin.com.intellij.psi.PsiElement
@@ -42,7 +41,7 @@ import org.jetbrains.kotlin.psi.psiUtil.containingClass
 import org.jetbrains.kotlin.psi.psiUtil.getChildOfType
 import org.jetbrains.kotlin.psi.psiUtil.parents
 import org.jetbrains.kotlin.psi.psiUtil.parentsWithSelf
-import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameSafe
+import org.jetbrains.kotlin.types.KotlinType
 import kotlin.reflect.KClass
 
 private val kotlinAnnotations = listOf(jvmSuppressWildcardsFqName, publishedApiFqName)
@@ -223,82 +222,6 @@ public fun KtClassOrObject.allPsiSuperClasses(
     ?: return emptyList()
 
   return listOf(superOrNull) + superOrNull.allPsiSuperClasses(module)
-}
-
-/**
- * Used to create a common type between [KtClassOrObject] class references and [ClassDescriptor]
- * references, to streamline parsing.
- *
- * @see allSuperTypeClassReferences
- */
-@ExperimentalAnvilApi
-public sealed interface ClassReference {
-
-  public val fqName: FqName
-
-  public data class Psi(
-    val clazz: KtClassOrObject,
-    override val fqName: FqName
-  ) : ClassReference
-
-  public data class Descriptor(
-    val clazz: ClassDescriptor,
-    override val fqName: FqName
-  ) : ClassReference
-}
-
-/**
- * This will return all super types as [ClassReference], whether they're parsed as [KtClassOrObject]
- * or [ClassDescriptor]. This will include generated code, assuming it has already been generated.
- * The returned sequence will be distinct by FqName, and Psi types are preferred over Descriptors.
- *
- * The first elements in the returned sequence represent the direct superclass to the receiver. The
- * last elements represent the types which are furthest up-stream.
- *
- * @param includeSelf If true, the receiver class is the first element of the sequence
- */
-@ExperimentalAnvilApi
-public fun KtClassOrObject.allSuperTypeClassReferences(
-  module: ModuleDescriptor,
-  includeSelf: Boolean = false
-): Sequence<ClassReference> {
-
-  fun KtClassOrObject.superTypeEntryNames() = superTypeListEntries
-    .map { it.requireFqName(module) }
-
-  fun FqName.typeRef(module: ModuleDescriptor): ClassReference {
-    return classDescriptorOrNull(module)
-      ?.let { Descriptor(it, this) }
-      ?: Psi(module.requireKtClassOrObject(this), this)
-  }
-
-  fun List<FqName>.typeRefs(module: ModuleDescriptor): Sequence<ClassReference> {
-    return asSequence().map { it.typeRef(module) }
-  }
-
-  val initial = if (includeSelf) {
-    listOf(requireFqName()) + superTypeEntryNames()
-  } else {
-    superTypeEntryNames()
-  }.typeRefs(module)
-
-  return generateSequence(initial) { superTypes ->
-    superTypes
-      .mapNotNull { classRef ->
-        when (classRef) {
-          is Psi -> classRef.clazz.superTypeEntryNames()
-            .takeIf { it.isNotEmpty() }
-            ?.typeRefs(module)
-          is Descriptor -> classRef.clazz.directSuperClassAndInterfaces()
-            .takeIf { it.isNotEmpty() }
-            ?.asSequence()
-            ?.map { it.fqNameSafe.typeRef(module) }
-        }
-      }
-      .flatten()
-      .takeIf { it.firstOrNull() != null }
-  }.flatten()
-    .distinctBy { it.fqName }
 }
 
 @ExperimentalAnvilApi
@@ -621,14 +544,54 @@ public fun KtClassOrObject.resolveTypeReference(
 ): KtTypeReference? {
 
   // if the element isn't a type variable name like `T`, it can be resolved through imports.
-  typeReference.typeElement?.fqNameOrNull(module)?.let { return typeReference }
+  typeReference.typeElement?.fqNameOrNull(module)
+    ?.let { return typeReference }
 
-  val declaringClass = typeReference.typeElement?.containingClass()
-    ?: throw AnvilCompilationException(
-      "Unable to find a containing class.",
-      element = typeReference
-    )
-  val declaringClassFqName = declaringClass.requireFqName()
+  val declaringClass = typeReference.requireContainingClassReference()
+  val parameterName = typeReference.text
+
+  return resolveGenericTypeReference(module, declaringClass, parameterName)
+}
+
+/**
+ * Safely resolves a [KotlinType], when that type reference is a generic expressed by a type
+ * variable name. This is done by inspecting the class hierarchy to find where the generic type is
+ * declared, then resolving *that* reference.
+ *
+ * For instance, given:
+ *
+ * ```
+ * interface SomeFactory : Function1<String, SomeClass>
+ * ```
+ *
+ * There's an invisible `fun invoke(p1: P1): R`. If `SomeFactory` is parsed using PSI (such as if
+ * it's generated), then the function can only be parsed to have the projected types of `P1` and `R`
+ *
+ * @receiver The class which actually references the type. In the above example, this would be
+ *   `SomeFactory`.
+ * @param declaringClass The class which declares the generic type name. In the above example, this
+ *   would be `Function1`.
+ * @param typeToResolve The generic reference to be resolved. In the above example, this is the `T`
+ *   **return type**.
+ */
+@ExperimentalAnvilApi
+public fun KtClassOrObject.resolveGenericKotlinType(
+  module: ModuleDescriptor,
+  declaringClass: ClassReference,
+  typeToResolve: KotlinType
+): KtTypeReference? {
+
+  val parameterName = typeToResolve.toString()
+
+  return resolveGenericTypeReference(module, declaringClass, parameterName)
+}
+
+private fun KtClassOrObject.resolveGenericTypeReference(
+  module: ModuleDescriptor,
+  declaringClass: ClassReference,
+  parameterName: String
+): KtTypeReference? {
+  val declaringClassFqName = declaringClass.fqName
 
   // If the class/interface declaring the generic is the receiver class,
   // then the generic hasn't been set to a concrete type and can't be resolved.
@@ -637,28 +600,56 @@ public fun KtClassOrObject.resolveTypeReference(
   }
 
   // Used to determine which parameter to look at in a KtTypeArgumentList
-  val indexToResolve = declaringClass.typeParameters
-    .indexOfFirst { it.requireIdentifier() == typeReference.typeElement?.text }
+  val indexOfType = declaringClass.indexOfTypeParameter(parameterName)
 
   // Find where the supertype is actually declared by matching the FqName of the SuperTypeListEntry
   // to the type which declares the generic we're trying to resolve.
   // After finding the SuperTypeListEntry, just take the TypeReference from its type argument list
-  val resolvedTypeReference = allSuperTypeClassReferences(module, includeSelf = true)
-    .filterIsInstance<Psi>()
-    .firstNotNullOfOrNull { typeRef ->
-      typeRef.clazz
-        .superTypeListEntries
-        .firstOrNull { it.requireFqName(module) == declaringClassFqName }
-    }
+  val resolvedTypeReference = superTypeListEntryOrNull(module, declaringClassFqName)
     ?.typeReference
     ?.typeElement
     ?.getChildOfType<KtTypeArgumentList>()
     ?.arguments
-    ?.get(indexToResolve)
+    ?.get(indexOfType)
     ?.typeReference
 
   return resolvedTypeReference?.takeIf { it.fqNameOrNull(module) != null }
     ?: resolvedTypeReference?.let { resolveTypeReference(module, it) }
+}
+
+/**
+ * Find where a super type is extended/implemented by matching the FqName of a SuperTypeListEntry to
+ * the FqName of the target super type.
+ *
+ * For instance, given:
+ *
+ * ```
+ * interface Base<T> {
+ *   fun create(): T
+ * }
+ *
+ * abstract class Middle : Comparable<MyClass>, Provider<Something>, Base<Something>
+ *
+ * class InjectClass : Middle()
+ * ```
+ *
+ * We start at the declaration of `InjectClass`, looking for a super declaration of `Base<___>`.
+ * Since `InjectClass` doesn't declare it, we look through the supers of `Middle` and find it, then
+ * resolve `T` to `Something`.
+ */
+@ExperimentalAnvilApi
+public fun KtClassOrObject.superTypeListEntryOrNull(
+  module: ModuleDescriptor,
+  superTypeFqName: FqName
+): KtSuperTypeListEntry? {
+  return toClassReference()
+    .allSuperTypeClassReferences(module, includeSelf = true)
+    .filterIsInstance<Psi>()
+    .firstNotNullOfOrNull { classReference ->
+      classReference.clazz
+        .superTypeListEntries
+        .firstOrNull { it.requireFqName(module) == superTypeFqName }
+    }
 }
 
 @ExperimentalAnvilApi
@@ -746,3 +737,19 @@ public fun KtClassOrObject.generateClassName(
         .shortName()
         .asString()
     }
+
+@ExperimentalAnvilApi
+public fun KtTypeReference.containingClassReferenceOrNull(): ClassReference? {
+  return typeElement
+    ?.containingClass()
+    ?.toClassReference()
+}
+
+@ExperimentalAnvilApi
+public fun KtTypeReference.requireContainingClassReference(): ClassReference {
+  return containingClassReferenceOrNull()
+    ?: throw AnvilCompilationException(
+      "Unable to find a containing class.",
+      element = this
+    )
+}

--- a/compiler/src/main/java/com/squareup/anvil/compiler/ClassScanner.kt
+++ b/compiler/src/main/java/com/squareup/anvil/compiler/ClassScanner.kt
@@ -2,7 +2,7 @@ package com.squareup.anvil.compiler
 
 import com.squareup.anvil.compiler.internal.annotationOrNull
 import com.squareup.anvil.compiler.internal.argumentType
-import com.squareup.anvil.compiler.internal.classDescriptorForType
+import com.squareup.anvil.compiler.internal.requireClassDescriptor
 import org.jetbrains.kotlin.descriptors.ClassDescriptor
 import org.jetbrains.kotlin.descriptors.ModuleDescriptor
 import org.jetbrains.kotlin.descriptors.PackageViewDescriptor
@@ -87,5 +87,5 @@ private class ContributedHint(properties: List<PropertyDescriptor>) {
   }
 
   private fun PropertyDescriptor.toClassDescriptor(): ClassDescriptor =
-    type.argumentType().classDescriptorForType()
+    type.argumentType().requireClassDescriptor()
 }

--- a/compiler/src/main/java/com/squareup/anvil/compiler/InterfaceMerger.kt
+++ b/compiler/src/main/java/com/squareup/anvil/compiler/InterfaceMerger.kt
@@ -3,9 +3,9 @@ package com.squareup.anvil.compiler
 import com.squareup.anvil.compiler.api.AnvilCompilationException
 import com.squareup.anvil.compiler.internal.annotationOrNull
 import com.squareup.anvil.compiler.internal.argumentType
-import com.squareup.anvil.compiler.internal.classDescriptorForType
 import com.squareup.anvil.compiler.internal.getAllSuperTypes
 import com.squareup.anvil.compiler.internal.getAnnotationValue
+import com.squareup.anvil.compiler.internal.requireClassDescriptor
 import com.squareup.anvil.compiler.internal.scope
 import org.jetbrains.kotlin.descriptors.ClassDescriptor
 import org.jetbrains.kotlin.descriptors.EffectiveVisibility.Public
@@ -123,7 +123,7 @@ internal class InterfaceMerger(
 
     val excludedClasses = (mergeAnnotation.getAnnotationValue("exclude") as? ArrayValue)
       ?.value
-      ?.map { it.argumentType(module).classDescriptorForType() }
+      ?.map { it.argumentType(module).requireClassDescriptor() }
       ?.filter { DescriptorUtils.isInterface(it) }
       ?.map { classDescriptorForExclusion ->
         val contributesToAnnotation = classDescriptorForExclusion

--- a/compiler/src/main/java/com/squareup/anvil/compiler/ModuleMerger.kt
+++ b/compiler/src/main/java/com/squareup/anvil/compiler/ModuleMerger.kt
@@ -8,8 +8,8 @@ import com.squareup.anvil.compiler.api.AnvilCompilationException
 import com.squareup.anvil.compiler.internal.annotation
 import com.squareup.anvil.compiler.internal.annotationOrNull
 import com.squareup.anvil.compiler.internal.argumentType
-import com.squareup.anvil.compiler.internal.classDescriptorForType
 import com.squareup.anvil.compiler.internal.getAnnotationValue
+import com.squareup.anvil.compiler.internal.requireClassDescriptor
 import com.squareup.anvil.compiler.internal.safePackageString
 import com.squareup.anvil.compiler.internal.scope
 import com.squareup.anvil.compiler.internal.toType
@@ -122,7 +122,7 @@ internal class ModuleMerger(
       ?.value
       ?.map {
         val argumentType = it.argumentType(module)
-        val classDescriptorForExclusion = argumentType.classDescriptorForType()
+        val classDescriptorForExclusion = argumentType.requireClassDescriptor()
 
         val contributesToAnnotation = classDescriptorForExclusion
           .annotationOrNull(contributesToFqName)

--- a/compiler/src/main/java/com/squareup/anvil/compiler/Utils.kt
+++ b/compiler/src/main/java/com/squareup/anvil/compiler/Utils.kt
@@ -10,9 +10,9 @@ import com.squareup.anvil.annotations.compat.MergeInterfaces
 import com.squareup.anvil.annotations.compat.MergeModules
 import com.squareup.anvil.compiler.api.AnvilCompilationException
 import com.squareup.anvil.compiler.internal.argumentType
-import com.squareup.anvil.compiler.internal.classDescriptorForType
 import com.squareup.anvil.compiler.internal.fqName
 import com.squareup.anvil.compiler.internal.getAnnotationValue
+import com.squareup.anvil.compiler.internal.requireClassDescriptor
 import dagger.Component
 import dagger.Lazy
 import dagger.MapKey
@@ -86,7 +86,7 @@ internal fun AnnotationDescriptor.boundType(
 ): ClassDescriptor {
   (getAnnotationValue("boundType") as? KClassValue)
     ?.argumentType(module)
-    ?.classDescriptorForType()
+    ?.requireClassDescriptor()
     ?.let { return it }
 
   val directSuperTypes = annotatedClass.getSuperInterfaces()
@@ -111,9 +111,7 @@ internal fun AnnotationDescriptor.boundType(
 internal fun AnnotationDescriptor.replaces(module: ModuleDescriptor): List<ClassDescriptor> {
   return (getAnnotationValue("replaces") as? ArrayValue)
     ?.value
-    ?.map {
-      it.argumentType(module).classDescriptorForType()
-    }
+    ?.map { it.argumentType(module).requireClassDescriptor() }
     ?: emptyList()
 }
 

--- a/compiler/src/main/java/com/squareup/anvil/compiler/codegen/BindingModuleGenerator.kt
+++ b/compiler/src/main/java/com/squareup/anvil/compiler/codegen/BindingModuleGenerator.kt
@@ -24,7 +24,6 @@ import com.squareup.anvil.compiler.internal.argumentType
 import com.squareup.anvil.compiler.internal.asClassName
 import com.squareup.anvil.compiler.internal.buildFile
 import com.squareup.anvil.compiler.internal.capitalize
-import com.squareup.anvil.compiler.internal.classDescriptorForType
 import com.squareup.anvil.compiler.internal.classesAndInnerClass
 import com.squareup.anvil.compiler.internal.decapitalize
 import com.squareup.anvil.compiler.internal.findAnnotation
@@ -546,7 +545,7 @@ private fun Binding.key(module: ModuleDescriptor): String {
         .map { (name, value) ->
           val valueString = when (value) {
             is KClassValue -> value.argumentType(module)
-              .classDescriptorForType().fqNameSafe.asString()
+              .requireClassDescriptor().fqNameSafe.asString()
             is EnumValue -> value.enumEntryName.asString()
             // String, int, long, ... other primitives.
             else -> value.toString()

--- a/compiler/src/main/java/com/squareup/anvil/compiler/codegen/DescriptorUtils.kt
+++ b/compiler/src/main/java/com/squareup/anvil/compiler/codegen/DescriptorUtils.kt
@@ -7,7 +7,6 @@ import com.squareup.anvil.compiler.internal.argumentType
 import com.squareup.anvil.compiler.internal.asClassName
 import com.squareup.anvil.compiler.internal.asTypeName
 import com.squareup.anvil.compiler.internal.capitalize
-import com.squareup.anvil.compiler.internal.classDescriptorForType
 import com.squareup.anvil.compiler.internal.isQualifier
 import com.squareup.anvil.compiler.internal.requireClassDescriptor
 import com.squareup.anvil.compiler.internal.toAnnotationSpec
@@ -83,7 +82,7 @@ fun PropertyDescriptor.isSetterInjected(): Boolean {
   return setter?.annotations?.hasAnnotation(injectFqName) == true
 }
 
-fun KotlinType.fqNameOrNull(): FqName? = classDescriptorForType()
+fun KotlinType.fqNameOrNull(): FqName? = requireClassDescriptor()
   .fqNameOrNull()
 
 internal fun List<CallableMemberDescriptor>.mapToMemberInjectParameters(

--- a/compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedFactoryGenerator.kt
+++ b/compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedFactoryGenerator.kt
@@ -20,8 +20,9 @@ import com.squareup.anvil.compiler.internal.allSuperTypeClassReferences
 import com.squareup.anvil.compiler.internal.asClassName
 import com.squareup.anvil.compiler.internal.asFunctionType
 import com.squareup.anvil.compiler.internal.asTypeName
+import com.squareup.anvil.compiler.internal.asTypeNameOrNull
 import com.squareup.anvil.compiler.internal.buildFile
-import com.squareup.anvil.compiler.internal.classDescriptorForType
+import com.squareup.anvil.compiler.internal.classDescriptorOrNull
 import com.squareup.anvil.compiler.internal.classesAndInnerClass
 import com.squareup.anvil.compiler.internal.findAnnotation
 import com.squareup.anvil.compiler.internal.findAnnotationArgument
@@ -33,8 +34,10 @@ import com.squareup.anvil.compiler.internal.hasAnnotation
 import com.squareup.anvil.compiler.internal.isInterface
 import com.squareup.anvil.compiler.internal.requireFqName
 import com.squareup.anvil.compiler.internal.requireTypeName
+import com.squareup.anvil.compiler.internal.resolveGenericKotlinType
 import com.squareup.anvil.compiler.internal.resolveTypeReference
 import com.squareup.anvil.compiler.internal.safePackageString
+import com.squareup.anvil.compiler.internal.toClassReference
 import com.squareup.anvil.compiler.internal.typeVariableNames
 import com.squareup.kotlinpoet.ClassName
 import com.squareup.kotlinpoet.FileSpec
@@ -49,6 +52,7 @@ import com.squareup.kotlinpoet.TypeSpec
 import com.squareup.kotlinpoet.asClassName
 import com.squareup.kotlinpoet.jvm.jvmStatic
 import dagger.internal.InstanceFactory
+import org.jetbrains.kotlin.descriptors.ClassDescriptor
 import org.jetbrains.kotlin.descriptors.DescriptorVisibilities.PROTECTED
 import org.jetbrains.kotlin.descriptors.DescriptorVisibilities.PUBLIC
 import org.jetbrains.kotlin.descriptors.FunctionDescriptor
@@ -285,28 +289,60 @@ internal class AssistedFactoryGenerator : PrivateCodeGenerator() {
   ) {
     companion object {
       fun FunctionDescriptor.toAssistedFactoryFunction(
+        module: ModuleDescriptor,
         factoryClass: KtClassOrObject
       ): AssistedFactoryFunction {
 
+        val containingClass = containingDeclaration as ClassDescriptor
+
         val returnTypeLazy = lazy(NONE) {
-          returnType
-            ?.classDescriptorForType()
-            ?.fqNameSafe
-            ?: throw AnvilCompilationException(
+
+          fun fail(): Nothing {
+            throw AnvilCompilationException(
               "Invalid return type: ${factoryClass.requireFqName()}. An assisted factory's " +
                 "abstract method must return a type with an @AssistedInject-annotated constructor.",
               element = factoryClass
             )
+          }
+
+          // This can be a generic type like `T` if the function is declared in a super type.
+          // Check that it can resolve before using this type.
+          returnType?.classDescriptorOrNull()
+            ?.fqNameSafe
+            ?: factoryClass.resolveGenericKotlinType(
+              module,
+              containingClass.toClassReference(),
+              returnType ?: fail()
+            )?.fqNameOrNull(module)
+            ?: fail()
         }
         val parameterPairs = lazy(NONE) {
           valueParameters.map { param ->
-            param.name.asString() to param.type.asTypeName()
+
+            val fromDescriptor = param.type
+              // This can be a generic type like `T` if the function is declared in a super type.
+              // Check that it can resolve before using this type.
+              .takeIf { it.classDescriptorOrNull() != null }
+              ?.asTypeName()
+              ?: factoryClass.resolveGenericKotlinType(
+                module,
+                containingClass.toClassReference(),
+                param.type
+              )?.requireTypeName(module)
+              ?: throw AnvilCompilationException(
+                "Unable to resolve assisted function's argument type: ${param.type}.",
+                element = factoryClass
+              )
+
+            param.name.asString() to fromDescriptor
           }
         }
         return AssistedFactoryFunction(
           name = name.asString(),
           returnTypeLazy = returnTypeLazy,
-          parameterKeys = lazy(NONE) { valueParameters.toKeysList() },
+          parameterKeys = lazy(NONE) {
+            valueParameters.toKeysList(module, containingClass, factoryClass)
+          },
           parameterPairs = parameterPairs
         )
       }
@@ -380,7 +416,7 @@ internal class AssistedFactoryGenerator : PrivateCodeGenerator() {
             .filterIsInstance<FunctionDescriptor>()
             .filter { it.modality == ABSTRACT }
             .filter { it.visibility == PUBLIC || it.visibility == PROTECTED }
-            .map { it.toAssistedFactoryFunction(this@requireSingleAbstractFunction) }
+            .map { it.toAssistedFactoryFunction(module, this@requireSingleAbstractFunction) }
             .toList()
         }
       }
@@ -389,7 +425,8 @@ internal class AssistedFactoryGenerator : PrivateCodeGenerator() {
     // `clazz` must be first in the list because of `distinctBy { ... }`, which keeps the first
     // matched element.  If the function's inherited, it can be overridden as well.  Prioritizing
     // the version from the file we're parsing ensures the correct variance of the referenced types.
-    val assistedFunctions = allSuperTypeClassReferences(module, includeSelf = true)
+    val assistedFunctions = toClassReference()
+      .allSuperTypeClassReferences(module, includeSelf = true)
       .distinctBy { it.fqName }
       .flatMap { it.assistedFunctions() }
       .distinctBy { it.name }
@@ -426,10 +463,25 @@ internal class AssistedFactoryGenerator : PrivateCodeGenerator() {
     val key: Int = identifier.hashCode() * 31 + typeName.hashCode()
 
     companion object {
-      fun Collection<ValueParameterDescriptor>.toKeysList(): List<AssistedParameterKey> =
+      fun Collection<ValueParameterDescriptor>.toKeysList(
+        module: ModuleDescriptor,
+        containingClass: ClassDescriptor,
+        factoryClass: KtClassOrObject
+      ): List<AssistedParameterKey> =
         map { descriptor ->
 
-          val type = descriptor.type.asTypeName()
+          val type = descriptor.type
+            .takeIf { it.classDescriptorOrNull() != null }
+            ?.asTypeNameOrNull()
+            ?: factoryClass.resolveGenericKotlinType(
+              module = module,
+              declaringClass = containingClass.toClassReference(),
+              typeToResolve = descriptor.type
+            )?.requireTypeName(module)
+            ?: throw AnvilCompilationException(
+              containingClass,
+              "Couldn't get type for parameter."
+            )
 
           AssistedParameterKey(
             typeName = type,

--- a/compiler/src/test/java/com/squareup/anvil/compiler/dagger/AssistedFactoryGeneratorTest.kt
+++ b/compiler/src/test/java/com/squareup/anvil/compiler/dagger/AssistedFactoryGeneratorTest.kt
@@ -482,6 +482,52 @@ public final class AssistedServiceFactory_Impl implements AssistedServiceFactory
     }
   }
 
+  @Test fun `the factory function may be provided by a generic super type with generic parameter`() {
+
+    // This is broken for Dagger, but working with Anvil factory generation
+    // https://github.com/google/dagger/issues/2984
+    if (useDagger) return
+
+    compile(
+      """
+      package com.squareup.test
+      
+      import dagger.assisted.Assisted
+      import dagger.assisted.AssistedFactory
+      import dagger.assisted.AssistedInject
+      import kotlin.properties.ReadOnlyProperty
+      import kotlin.reflect.KProperty
+      
+      data class AssistedService @AssistedInject constructor(
+          val int: Int,
+          @Assisted val strings: List<String>
+      )
+      
+      @AssistedFactory
+      interface AssistedServiceFactory : Function1<List<String>, AssistedService> 
+      """
+    ) {
+      val factoryImplClass = assistedServiceFactory.implClass()
+      val generatedFactoryInstance = assistedService.factoryClass().createInstance(Provider { 5 })
+      val factoryImplInstance = factoryImplClass.createInstance(generatedFactoryInstance)
+
+      val staticMethods = factoryImplClass.declaredMethods.filter { it.isStatic }
+      assertThat(staticMethods).hasSize(1)
+
+      val factoryProvider = staticMethods.single { it.name == "create" }
+        .invoke(null, generatedFactoryInstance) as Provider<*>
+      assertThat(factoryProvider.get()::class.java).isEqualTo(factoryImplClass)
+
+      val assistedServiceInstance = factoryImplClass.declaredMethods
+        .filterNot { it.isStatic }
+        .last { it.name == "invoke" }
+        .invoke(factoryImplInstance, listOf("a"))
+
+      assertThat(assistedServiceInstance)
+        .isEqualTo(assistedService.createInstance(5, listOf("a")))
+    }
+  }
+
   @Test fun `the implementation function name matches the factory name`() {
     compile(
       """

--- a/integration-tests/code-generator/src/main/java/com/squareup/anvil/test/TestCodeGenerator.kt
+++ b/integration-tests/code-generator/src/main/java/com/squareup/anvil/test/TestCodeGenerator.kt
@@ -102,14 +102,10 @@ class TestCodeGenerator : CodeGenerator {
             @Assisted val string: String
           )
 
-          interface ParentAssistedFactory {
-            fun create(string: String): AssistedService
-          }
+          interface ParentAssistedFactory : Function1<String, AssistedService>
     
           @AssistedFactory
-          interface SampleAssistedFactory : ParentAssistedFactory {
-            override fun create(string: String): AssistedService
-          }
+          interface SampleAssistedFactory : ParentAssistedFactory
         """.trimIndent()
 
         sequenceOf(

